File: SudokuSolver.rb

class Sudoku

    attr_reader :path, :puzzle, :n, :num_guesses

    def initialize path
        @path = path
        @n = 0
        @b = 0
        @puzzle = []
        @num_guesses = 0
        @known_indices = []
        load_puzzle
    end

    def to_s
        s = ''
        current_row = 0
        @puzzle.each_with_index do |c, i|
            r = (i / @n).floor
            s << '\n' if r != current_row
            s << c.to_s
            current_row = r
        end
        return s
    end

    def solve!
        @num_guesses = 0
        r = solve_from! 0, 1
        while r != nil
            r = solve_from! r[0], r[1]
        end
    end

    private

    def solve_from! index, starting_guess
        if index < 0 or index > @puzzle.length
            puts to_s.split '\n'
            raise "Invalid puzzle index #{index} after #{@num_guesses} guesses"
        end

        last_valid_guess_index = nil
        found_valid_guess = false
        (index...@puzzle.length).each do |i|
            unless @known_indices.include? i
                found_valid_guess = false
                (starting_guess..@n).each do |guess|
                    @num_guesses += 1
                    if valid? i, guess
                        found_valid_guess = true
                        last_valid_guess_index = i
                        @puzzle[i] = guess
                        break
                    end
                end
                starting_guess = 1

                if not found_valid_guess
                    break
                end
            end
        end

        if not found_valid_guess
            new_index = index - 1
            new_index = last_valid_guess_index unless last_valid_guess_index == nil
            new_starting_guess = @puzzle[new_index] + 1

            reset_puzzle_at new_index

            while new_starting_guess > @n or @known_indices.include? new_index
                new_index -= 1
                new_starting_guess = @puzzle[new_index] + 1
                reset_puzzle_at new_index
            end

            return [new_index, new_starting_guess]
        else
            return nil
        end
    end

    def reset_puzzle_at index
        (index...@puzzle.length).each do |i|
            unless @known_indices.include? i
                @puzzle[i] = 0
            end
        end
    end

    def valid? index, guess
        # check column
        col_index = index % @n
        (0...@n).each do |r|
            r_index = col_index + (@n * r)
            unless r_index == index
                if @puzzle[r_index] == guess
                    return false
                end
            end
        end

        # check row
        row_index = (index / @n).floor
        start = @n * row_index
        finish = start + @n
        (start...finish).each do |c_index|
            unless c_index == index
                if @puzzle[c_index] == guess
                    return false
                end
            end
        end

        # check block
        return valid_for_block? index, guess
    end

    def valid_for_block? index, guess
        row_index = (index / @n).floor
        col_index = index % @n

        block_row = (row_index / @b).floor
        block_col = (col_index / @b).floor

        row_start = block_row * @b
        row_end = row_start + @b - 1
        col_start = block_col * @b
        col_end = col_start + @b - 1

        (row_start..row_end).each do |r|
            (col_start..col_end).each do |c|
                i = c + (r * @n)
                if @puzzle[i] == guess
                    return false
                end
            end
        end

        return true
    end

    def load_puzzle
        @n = 0
        @puzzle = []
        File::open(@path, 'r') do |file|
            i = 0
            while content = file.gets
                @n = content.to_s.length - 1 if @n == 0
                content.to_s.each_char do |c|
                    if c =~ /\d/
                        @puzzle << c.to_i
                        @known_indices << i if c.to_i != 0
                        i += 1
                    end
                end
            end
        end
        @b = Math.sqrt(@n).to_i
    end

end


if __FILE__ == $0
    s = Sudoku.new ARGV[0]

    puts
    puts 'Puzzle:'
    puts s.to_s.split '\n'
    puts

    puts 'Solution:'
    s.solve!
    puts s.to_s.split '\n'
    puts "Number of guesses: #{s.num_guesses}"
    puts
end

File: SudokuSolver.py
import math
from datetime import datetime

class Sudoku(object):

    def __init__(self):
        self.n = 0
        self.b = 0
        self.puzzle = []
        self.num_guesses = 0
        self.known_indices = []

    def load_puzzle(self, puzzle_string):
        self.puzzle = []
        self.known_indices = []
        rows = [row for row in puzzle_string.split("\n") if row]
        self.n = len(rows)
        self.b = int(math.sqrt(self.n))
        for row_index, row in enumerate(rows):
            for col_index, c in enumerate(row):
                if c.isdigit():
                    self.puzzle.append(int(c))
                    if int(c) != 0:
                        self.known_indices.append((row_index * self.n) + col_index)

    def load_puzzle_from_file(self, path):
        with open(path, 'r') as f:
            self.load_puzzle(f.read())

    def to_string(self, pretty=True):
        return "".join(['\n' + str(c) if i % self.n == 0 else str(c) for i, c in enumerate(self.puzzle)])

    def __repl__(self):
        return self.to_string(pretty=False)

    def __str__(self):
        return self.to_string()

    def solve(self):
        start_time = datetime.now()
        self.num_guesses = 0
        r = self.solve_from(0, 1)
        while r is not None:
            r = self.solve_from(r[0], r[1])
        return datetime.now() - start_time

    def solve_from(self, index, starting_guess):
        if index < 0 or index > len(self.puzzle):
            raise Exception("Invalid puzzle index %s after %s guesses" % (index, self.num_guesses))

        last_valid_guess_index = None
        found_valid_guess = False
        for i in xrange(index, len(self.puzzle)):
            if i not in self.known_indices:
                found_valid_guess = False
                for guess in xrange(starting_guess, self.n + 1):
                    self.num_guesses += 1
                    if self.valid(i, guess):
                        found_valid_guess = True
                        last_valid_guess_index = i
                        self.puzzle[i] = guess
                        break

                starting_guess = 1
                if not found_valid_guess:
                    break
        
        if not found_valid_guess:
            new_index = last_valid_guess_index if last_valid_guess_index is not None else index - 1
            new_starting_guess = self.puzzle[new_index] + 1
            self.reset_puzzle_at(new_index)

            while new_starting_guess > self.n or new_index in self.known_indices:
                new_index -= 1
                new_starting_guess = self.puzzle[new_index] + 1
                self.reset_puzzle_at(new_index)
            
            return (new_index, new_starting_guess)
        else:
            return None

    def reset_puzzle_at(self, index):
        for i in xrange(index, len(self.puzzle)):
            if i not in self.known_indices:
                self.puzzle[i] = 0

    def valid_for_row(self, index, guess):
        row_index = int(math.floor(index / self.n))
        start = self.n * row_index
        finish = start + self.n
        for c_index in xrange(start, finish):
            if c_index != index and self.puzzle[c_index] == guess:
                return False
        return True

    def valid_for_column(self, index, guess):
        col_index = index % self.n
        for r in xrange(0, self.n):
            r_index = col_index + (self.n * r)
            if r_index != index and self.puzzle[r_index] == guess:
                return False
        return True

    def valid_for_block(self, index, guess):
        row_index = int(math.floor(index / self.n))
        col_index = index % self.n

        block_row = int(math.floor(row_index / self.b))
        block_col = int(math.floor(col_index / self.b))

        row_start = block_row * self.b
        row_end = row_start + self.b - 1
        col_start = block_col * self.b
        col_end = col_start + self.b - 1

        for r in xrange(row_start, row_end + 1):
            for c in xrange(col_start, col_end + 1):
                i = c + (r * self.n)
                if self.puzzle[i] == guess:
                    return False

        return True

    def valid(self, index, guess):
        return self.valid_for_row(index, guess) and self.valid_for_column(index, guess) and self.valid_for_block(index, guess)


if __name__ == "__main__":
    import sys

    s = Sudoku()
    s.load_puzzle_from_file(sys.argv[1])

    print "\nPuzzle:\n%s" % (s)

    time = s.solve()
    print "\nSolution:\n%s" % (s)
    print "\nNumber of guesses: %s" % (s.num_guesses)
    print "Total time: %s" % (time) 

